### PR TITLE
Improve flexibility of the `BoundenContext` creation

### PR DESCRIPTION
**Pull request by dmytro-grankin**
This PR improves the flexibility of the `BoundedContext` creation, allowing to specify a custom `StorageFactory`. Also the factory for bounded contexts was renamed from `TodoListBounedContext` to `BoundedContexts`.

### CODE REVIEW

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java [hunk 1]
```diff
@@ -33,55 +33,54 @@
...
+import static io.spine.util.Exceptions.newIllegalStateException;
+
 /**
- * Serves for creation the {@link BoundedContext} instances.
+ * Utilities for creation the {@link BoundedContext} instances.
  *
  * @author Illia Shepilov
  */
-public class TodoListBoundedContext {
+public class BoundedContexts {
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java]: Looks `final`.

File: cli-app/src/main/java/io/spine/examples/todolist/TodoList.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import static io.spine.util.Exceptions.illegalStateWithCauseOf;
+
+/**
+ * A facade for the application execution.
+ *
+ * @author Dmytro Grankin
+ */
+public final class TodoList {
+
+    private final Server server;
```
Comment (armiol) [inline cli-app/src/main/java/io/spine/examples/todolist/TodoList.java]: When does it ever runs like that? `Server` and `TodoClient` in a single class do not make any sense.

Let's discuss your intention.

File: server/src/main/java/io/spine/examples/todolist/server/Server.java [hunk 3]
```diff
@@ -100,13 +98,9 @@ private void awaitTermination() {
     }
 
     /**
-     * Initiates an orderly shutdown of {@link GrpcContainer} and closes {@link BoundedContext}.
-     *
-     * <p> Closes the {@code BoundedContext} performing all necessary clean-ups.
-     *
-     * @throws Exception caused by closing one of the {@link BoundedContext} components
+     * Initiates an orderly shutdown of the {@link GrpcContainer}.
```
Comment (armiol) [inline server/src/main/java/io/spine/examples/todolist/server/Server.java]: Let's simplify that to `Initiates a shutdown of this {@code Server} instance.`

File: cli-app/src/main/java/io/spine/examples/todolist/view/NewTaskView.java [hunk 4]
```diff
@@ -122,13 +125,17 @@ public void start(Screen screen, CreateBasicTaskVBuilder builder) {
     private static class CreateTask extends CommandAction<CreateBasicTask,
                                                           CreateBasicTaskVBuilder> {
 
-        private CreateTask(CommandView<CreateBasicTask, CreateBasicTaskVBuilder> source) {
+        private final TodoClient client;
```
Comment (armiol) [inline cli-app/src/main/java/io/spine/examples/todolist/view/NewTaskView.java]: I don't understand this as well.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java (hunk 1 above)
Comment (dmytro-grankin) [reply api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java]: Done.

File: cli-app/src/main/java/io/spine/examples/todolist/TodoList.java (hunk 2 above)
Comment (dmytro-grankin) [reply cli-app/src/main/java/io/spine/examples/todolist/TodoList.java]: Done.

File: server/src/main/java/io/spine/examples/todolist/server/Server.java (hunk 3 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/examples/todolist/server/Server.java]: Done.

File: cli-app/src/main/java/io/spine/examples/todolist/view/NewTaskView.java (hunk 4 above)
Comment (dmytro-grankin) [reply cli-app/src/main/java/io/spine/examples/todolist/view/NewTaskView.java]: Done.

**Comment by dmytro-grankin** [comment]
@armiol PTAL.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java [hunk 5]
```diff
@@ -33,55 +33,54 @@
...
-    private static final StorageFactory storageFactory =
+    private static final StorageFactory IN_MEMORY_FACTORY =
             InMemoryStorageFactory.newInstance(NAME, false);
 
-    /**
-     * Obtains the reference to the singleton {@link BoundedContext}.
-     */
-    public static BoundedContext getInstance() {
-        return Singleton.INSTANCE.value;
+    private BoundedContexts() {
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java]: Have a test for the private ctor.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java [hunk 6]
```diff
@@ -33,55 +33,54 @@
...
 import io.spine.server.storage.memory.InMemoryStorageFactory;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.spine.util.Exceptions.newIllegalStateException;
+
 /**
- * Serves for creation the {@link BoundedContext} instances.
+ * Utilities for creation the {@link BoundedContext} instances.
  *
  * @author Illia Shepilov
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java]: You are the author as well.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java [hunk 7]
```diff
@@ -111,19 +111,18 @@ private static BoundedContext create() {
         return eventBus;
     }
 
+    @SuppressWarnings("Guava" /* Spine API is Java 7-based
+                                 and uses `Optional` from Google Guava. */)
     private static BoundedContext createBoundedContext(EventBus.Builder eventBus) {
+        final Optional<StorageFactory> storageFactory = eventBus.getStorageFactory();
+        if (!storageFactory.isPresent()) {
+            throw newIllegalStateException("EventBus does not specify a StorageFactory.");
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContexts.java]: This isn't covered.

**Pull request merged**